zsh-syntax-highlighting.zsh

```diff
@@ -27,571 +27,9 @@
-       type "_zsh_highlight_highlighter_${highlighter}_predicate" &> /dev/null;
-    then
-        # New (0.5.0) function names
-    elif type "_zsh_highlight_${highlighter}_highlighter" &> /dev/null &&
-         type "_zsh_highlight_${highlighter}_highlighter_predicate" &> /dev/null;
-    then
-        # Old (0.4.x) function names
-        if false; then
-            # TODO: only show this warning for plugin authors/maintainers, not for end users
-            print -r -- >&2 "zsh-syntax-highlighting: warning: ${(qq)highlighter} highlighter uses deprecated entry point names; please ask its maintainer to update it: https://github.com/zsh-users/zsh-syntax-highlighting/issues/329"
-        fi
-        # Make it work.
-        eval "_zsh_highlight_highlighter_${(q)highlighter}_paint() { _zsh_highlight_${(q)highlighter}_highlighter \"\$@\" }"
-        eval "_zsh_highlight_highlighter_${(q)highlighter}_predicate() { _zsh_highlight_${(q)highlighter}_highlighter_predicate \"\$@\" }"
-    else
-        print -r -- >&2 "zsh-syntax-highlighting: ${(qq)highlighter} highlighter should define both required functions '_zsh_highlight_highlighter_${highlighter}_paint' and '_zsh_highlight_highlighter_${highlighter}_predicate' in ${(qq):-"$highlighter_dir${highlighter}-highlighter.zsh"}."
-    fi
-  done
-}
-
-
-# -------------------------------------------------------------------------------------------------
-# Setup
-# -------------------------------------------------------------------------------------------------
-
-# Try binding widgets.
-_zsh_highlight_bind_widgets || {
-  print -r -- >&2 'zsh-syntax-highlighting: failed binding ZLE widgets, exiting.'
-  return 1
-}
-
-# Resolve highlighters directory location.
-_zsh_highlight_load_highlighters "${ZSH_HIGHLIGHT_HIGHLIGHTERS_DIR:-${${0:A}:h}/highlighters}" || {
-  print -r -- >&2 'zsh-syntax-highlighting: failed loading highlighters, exiting.'
-  return 1
-}
-
-# Reset scratch variables when commandline is done.
-_zsh_highlight_preexec_hook()
-{
-  typeset -g _ZSH_HIGHLIGHT_PRIOR_BUFFER=
-  typeset -gi _ZSH_HIGHLIGHT_PRIOR_CURSOR=
-}
-autoload -Uz add-zsh-hook
-add-zsh-hook preexec _zsh_highlight_preexec_hook 2>/dev/null || {
-    print -r -- >&2 'zsh-syntax-highlighting: failed loading add-zsh-hook.'
-  }
-
-# Load zsh/parameter module if available
-zmodload zsh/parameter 2>/dev/null || true
-
-# Initialize the array of active highlighters if needed.
-[[ $#ZSH_HIGHLIGHT_HIGHLIGHTERS -eq 0 ]] && ZSH_HIGHLIGHT_HIGHLIGHTERS=(main)
-
-if (( $+X_ZSH_HIGHLIGHT_DIRS_BLACKLIST )); then
-  print >&2 'zsh-syntax-highlighting: X_ZSH_HIGHLIGHT_DIRS_BLACKLIST is deprecated. Please use ZSH_HIGHLIGHT_DIRS_BLACKLIST.'
-  ZSH_HIGHLIGHT_DIRS_BLACKLIST=($X_ZSH_HIGHLIGHT_DIRS_BLACKLIST)
-  unset X_ZSH_HIGHLIGHT_DIRS_BLACKLIST
-fi
-
-# Restore the aliases we unned
-eval "$zsh_highlight__aliases"
-builtin unset zsh_highlight__aliases
+emulate zsh -c 'source ${0:h}/driver.zsh'
```

Comment:
An interim commit used `0=${(%):-%N}; fpath+=( ${0:a:h} )`.  For the record, I'm not sure that interim state was correct in all cases, because `:a`, unlike `:P`, is not guaranteed to return something that refers to the same inode (or even to _an_ inode).  I think the compatible way to do this would have been to use ${0:P:h} if available, and `$(cd -q -- ${0:h} && pwd -P)` otherwise.

But that was just an interim state, and I'm mentioning this just for completeness.  I'm not suggesting to rewrite the history over this.

Reply:
`fpath+=(${0:a:h})` is still used just in driver.zsh now. Some resolution needs to happen to avoid a relative path (which causes the tests to fail).

Do any of the pathological `:a` cases apply here with `$0`?

Reply:
I think so; watch:

```sh
% file /tmp/{slash,foo,foo/bar}
/tmp/slash:   symbolic link to /
/tmp/foo:     directory
/tmp/foo/bar: Paul Falstad's zsh script, ASCII text executable
% cat /tmp/slash/../tmp/foo/bar
#!/bin/zsh -f
echo "[$0]" I am /tmp/foo/bar
% source /tmp/slash/../tmp/foo/bar 
[/tmp/slash/../tmp/foo/bar] I am /tmp/foo/bar
% echo ${${:-/tmp/slash/../tmp/foo/bar}:a}
/tmp/tmp/foo/bar
% 
```

Reply:
Fixed to use `:P`.